### ISSUE TITLE
modify runner 

### ISSUE THREAD

**Issue by varde80** [varde80/aimat#3]
runner 클래스 등을 수정하여 우리가 지정한 메트릭을 yaml 파일에서 지정하여 사용할수 있도록 하여주세요

해당 메트릭등은 Metric class의 서브 클래스이거나 이를 수정한 형태로 개발합니다. 

1차적으로 아이디어를 여기에 적은 후 구현 바랍니다.

**Comment by ckChef** [comment]
> runner 클래스 등을 수정하여 우리가 지정한 메트릭을 yaml 파일에서 지정하여 사용할수 있도록 하여주세요
> 
> 해당 메트릭등은 Metric class의 서브 클래스이거나 이를 수정한 형태로 개발합니다.
> 
> 1차적으로 아이디어를 여기에 적은 후 구현 바랍니다.

@varde80 박사님, 말씀해 주신 issue 에 대해 현재 구상중인 구현방안 말씀 드립니다.
# 1. metric 지정 방법
필요한 metrics 종류의 이름(str type)이 담긴 list를 yaml 파일에서 지정하고, 이를 Runner의 생성자로 전달해 주면 좋을것 같습니다.

현재 legacy code와 비교한 예시는 아래와 같습니다.
legacy
```python3
class Runner:
    def __init__(
        self,
        loader: DataLoader[Any],
        model: torch.nn.Module,
        criterion: torch.nn.Module,
        device: str,
        optimizer: Optional[torch.optim.Optimizer] = None,
    ) -> None:
```

new
```python3
class Runner:
    def __init__(
        self,
        loader: DataLoader[Any],
        model: torch.nn.Module,
        criterion: torch.nn.Module,
        device: str,
        optimizer: Optional[torch.optim.Optimizer] = None,
        metrics: list,
    ) -> None:
```

## list type으로 전달하는 이유
1. 추상화: 경우에 따라 metric의 종류나 그 개수가 각기 다를 수 있기 때문입니다.
2. 코드작성 효율성 향상: list에 담긴 element (str type의 metric 이름) 재사용 가능
    - 만약 결과를 dictionary로 저장한다면, list에 담긴 metric_name 을 그대로 dictionary의 key로 사용할 수 있습니다.
    - tensorboard의 writer.add_scalar()와 같이 tag를 전달해 줘야할 때도 그대로 사용할 수 있습니다.

## user가 사용하는 경우 예시
https://github.com/varde80/aimat/blob/452055fd8f1b93c96439a80230be062a3ecef57b/conf/config.yaml#L1-L20
config.yaml 파일에 계산 필요한 metric 종류를 지정하는 경우, 아래와 같은 내용을 추가하면 됩니다.
```yaml
metrics:
- 'r2'
- 'ssim'
- 'psnr'
```
그러고 나서 main.py에서 cfg를 print해보면 list type의 metrics가 전달 된 것을 확인할 수 있습니다.
```console
{'batch_size': 64,
 'batch_size_test': 16,
 'criterion': {'_target_': 'torch.nn.CrossEntropyLoss'},
 'dataset': {'dataset': 'CIFAR100'},
 'device': 'mps',
 'iscustom': False,
 'logger': 'tb',
 'lr': 0.001,
 'max_epochs': 10,
 'metrics': ['r2', 'ssim', 'psnr'],
 'model': {'_target_': 'aimat.models.CIFAR100_ConvNet'},
 'optimizer': {'_target_': 'torch.optim.RMSprop', 'lr': '${lr}', 'alpha': 0.99, 'eps': 1e-08, 'weight_decay': 0, 'momentum': 0, 'centered': False},
 'paths': {'data_dir': '${hydra:runtime.cwd}/data/', 'log_dir': '${hydra:runtime.cwd}/runs/', 'save_dir': '${hydra:runtime.cwd}/ckpt/'},
 'project_name': 'varde-mnist',
 'save_model': True,
 'save_model_interval': 1,
 'scheduler': {'_target_': 'torch.optim.lr_scheduler.ReduceLROnPlateau', 'mode': 'min', 'factor': 0.1, 'patience': 10, 'threshold': 0.0001, 'threshold_mode': 'rel', 'cooldown': 0, 'min_lr': 0, 'eps': 1e-08, 'verbose': False}}
```

## Runner의 생성자로 전달하는 이유
1. Runner 내부 method에서 학습이 이루어 지고, 학습 결과로 부터 metric을 계산할 때 어느 metric을 계산해야할지 알기위해 metrics가 필요하기 때문입니다. 
2. 일반적으로 학습하는 동안 필요한 metric의 종류가 변하는 경우가 없기 때문에, 처음 Runner를 생성할 때 지정하는 __init__으로 전달하는게 적절하다 생각했습니다.

# 2. Metric class 구현 방안
> 해당 메트릭등은 Metric class의 서브 클래스이거나 이를 수정한 형태로 개발합니다.

issue에서 말씀하신 것 처럼, Metric으로 부터 상속 받아 subclass 형태로 구현하고자 합니다

단 MetricCaller라는 추가적인 class를 정의해서 사용하는 방법을 구상하고 있으며

전체적인 컨셉을 나타낸 예시 코드는 아래와 같습니다.

예시를 위해 aimat.core.metric.py를 수정하는 방법으로 작성 하였으며 MAE와 PSNR을 계산해야하는 경우 입니다.
```python3
from dataclasses import dataclass, field
import numpy as np
import torch
from skimage.metrics import peak_signal_noise_ratio as psnr


@dataclass
class Metric:
    values: list[float] = field(default_factory=list)
    running_total: float = 0.0
    num_updates: float = 0.0
    average: float = 0.0

    def update(self, value: float, batch_size: int):
        self.values.append(value)
        self.running_total += value * batch_size
        self.num_updates += batch_size
        self.average = self.running_total / self.num_updates

    def calc(self):
        raise NotImplementedError


class MetricMAE(Metric):
    is_np = False  # NOTE 2: metric 계산 위해 np.ndarray 객체일 필요 없는 경우

    def calc(self, pred, gt, n_batch):
        rst = (pred - gt).abs().mean().item()

        self.update(rst, n_batch)


class MetricPSNR(Metric):
    is_np = True  # NOTE 2: metric 계산 위해 np.ndarray 객체여야 하는 경우

    # NOTE: psnr()은 torch.tensor 객체 사용이 불가능하며, batch 단위 계산이 불가능함
    def calc(self, pred, gt, n_batch):
        rst = np.zeros(n_batch, dtype=np.float32)
        for idx_batch in range(n_batch):
            rst[idx_batch] = psnr(gt[idx_batch], pred[idx_batch], data_range=2)

        rst = rst.mean()  # batch mean

        self.update(rst, n_batch)


class MetricCaller:  # Runner 내의 학습 코드 안에서 instance가 생성되고 calc_metrics call 됨
    def __init__(self, name_list: list) -> None:
        # NOTE 1: metric 계산에 np.ndarray type 필요 없는 경우, 불필요하게 tensor -> np로 변환하는것 방지하기 위해 미리 type casting 안함
        # NOTE 1-1: type casting 과정이 overhead 유발하므로, 한 번 하고나면 저장해서 재사용
        self.pred_np = None
        self.gt_np = None

        self.name_list = name_list
        self.result_total = {
            # HACK: 각 metric 이름(key)에 해당하는 Metric class(value)를 manual하게 지정해야함
            'mae': MetricMAE(),
            'psnr': MetricPSNR(),
        }

        for name_metric in self.name_list:  # name_list element 처리할 수 있는지 검사
            assert name_metric in self.result_total

    def flush_data(self):  # 다음 batch data 넘어가기 전에 call 해줘야 하는 method
        self.pred_np = None
        self.gt_np = None

    def calc_metrics(self, pred: torch.Tensor, gt: torch.Tensor, n_batch):
        for name_metric in self.name_list:
            # NOTE 1: 여기서 metric에 저장된 is_np 확인해보고 np.ndarray 필요하면 할당
            if self.result_total[name_metric].is_np:  # NOTE 2: (pred, gt)가 numpy.ndarray 객체여야 계산 가능한 경우
                if self.pred_np is None:  # NOTE 1-1
                    self.pred_np = pred.clone().cpu().detach().numpy()
                if self.gt_np is None:  # NOTE 1-1
                    self.gt_np = gt.clone().cpu().detach().numpy()

                self.result_total[name_metric].calc(self.pred_np, self.gt_np, n_batch)

            else:
                self.result_total[name_metric].calc(pred, gt, n_batch)

        self.flush_data()


if __name__ == '__main__':
    metrics_list = ['mae', 'psnr']  # 실제론 yaml로 부터, cfg를 통해 전달될 객체

    mc = MetricCaller(metrics_list)
    ''' user가 이 metric.py에서 정의되지 않은 custom class를 사용하고 싶은 경우 아래와 같이 사용
    mc.result_total[name_custom] = calc_metric_object
    '''
    # batch iteration loop 안이라고 가정 ############################################################
    # NOTE: 매 batch interation 마다 batch dimension의 크기가 변하는 상황 가정
    for idx_batch in range(2, 4):  # 원래 for data, label in dataloader: 가 있는 자리
        pred = torch.randn((idx_batch, 3, 64, 64)).to('cuda:0')  # gpu로 계산 끝낸 직후라고 가정
        gt = torch.randn((idx_batch, 3, 64, 64)).to('cuda:0')
        mc.calc_metrics(pred, gt, pred.size(0))
    ################################################################################################

    # NOTE: average 값 만 있는게 아니라 Metric의 다른 variable들(values, running_total, ...) 아직 있는 상태
    print(mc.result_total)

```
위 코드를 실행한 결과는 아래와 같습니다.
```console
{'mae': MetricMAE(values=[1.1335241794586182, 1.1241003274917603], running_total=5.639349341392517, num_updates=5.0, average=1.1278698682785033), 'psnr': MetricPSNR(values=[2.9979918, 3.0130663], running_total=15.035182476043701, num_updates=5.0, average=3.00703649520874)}
```

## class MetricCaller
`def __init__` params
- name_list: 계산해야될 metric 이름이 저장된 list입니다.

`def calc_metrics` params
- pred, gt: tensor.Tensor인 data들 입니다.
- n_batch: int type으로 data의 batch size를 알려 줍니다. 이게 필요한 경우는 크게 두가지인데 먼저 Metric class의 update method가 call 될 때, 그리고 batch 단위의 계산이 불가능 한 metric을 계산할 때 입니다.

별도 return은 없고 해당 클래스의 인스턴스로 부터 바로 mc.result_total 형태로 결과 dictionary를 참조할 수 있습니다.

### MetricCaller 사용 이유
- 추상화: 각 metric 별로 계산에 필요한 계산 class가 서로다르므로 적절한 계산 class를 연결하는 과정이 요구됨
- 최적화: 기본적으로 torch.Tensor에서 numpy.ndarray 객채로 casting 하는 것은 overhead를 유발
  - 이미 구현돼 있는 metric 계산 함수(e.g. skimage.metric)를 사용할 때 torch.Tensor type사용이 불가능해서 불가피하게 type casting이 요구되는 경우가 있음
  - numpy.ndarray type 을 요구하는 metric이 여러개인 경우, 한 번 type casting 하고나면 재사용 하는 것이 효율적이므로
  - 동시에 numpy.ndarray가 필요없는 metric만 계산해야 한다면, 아예 type casting을 하지 않는게 효율적이므로

---
구현방안은 이상입니다.

감사합니다.

**Comment by varde80** [comment]
아주 좋습니다. 해당브랜치에서 진행하시기 바랍니다.

**Comment by ckChef** [comment]
@varde80, @namKolorfuL

현재 구현 완료된 사용가능 Metric 종류
- MAE
- MSE
- RMSE
- SSIM
- PSNR
- Accuracy

구현 예정 Metric
- R2 score
- f1 score
- IOU

위 목록들의 Metric들이 코드에 추가될 예정입니다.

혹시 저 Metric 들 외에 평소에 자주 사용하셔서 필요한 기능이 있으시다면 말씀 부탁드립니다.
